lib/wallaroo/core/sink/connector_sink/connector_sink_notify.pony

```diff
@@ -0,0 +1,455 @@
+    @printf[I32]("ConnectorSink connection failed\n".cstring())
+
+  fun ref expect(conn: WallarooOutgoingNetworkActor ref, qty: USize): USize =>
+    qty
+
+  fun ref received(conn: WallarooOutgoingNetworkActor ref, data: Array[U8] iso,
+    times: USize): Bool
+  =>
+    if _header then
+      try
+        let payload_size: USize = _payload_length(consume data)?
+
+        conn.expect(payload_size)
+        _header = false
+      else
+        Fail()
+      end
+      true
+    else
+      conn.expect(4)
+      _header = true
+      let data' = recover val consume data end
+      try
+        _process_connector_sink_v2_data(conn, data')?
+      else
+        Fail()
+      end
+      true
+    end
+
+  fun ref sent(conn: WallarooOutgoingNetworkActor ref, data: (String val | Array[U8 val] val))
+    : (String val | Array[U8 val] val)
+  =>
+    """
+    All 2PC messages are sent via this class's send_msg(), which uses
+    conn._write_final(), which does *not* filter its data through the
+    sent() callback.
+    Also, all Wallaroo data that is buffered in twopc_reconnect_buffer
+    and later sent after we are unthrottled uses conn._write_final().
+
+    As a result, this function should be unreachable.
+    """
+    Unreachable()
+    data
+
+  fun ref sentv(conn: WallarooOutgoingNetworkActor ref,
+    data: ByteSeqIter): ByteSeqIter
+  =>
+    """
+    All non-2PC data is written to the ConnectorSink via writev(),
+    thus all non-2PC data (i.e., all Wallaroo app data that is sent
+    to the sink) is filtered by this callback.
+    """
+    if _connected and (not _throttled) then
+      data
+    else
+      ifdef "checkpoint_trace" then
+        @printf[I32]("Sink sentv: not connected or throttled: buffering\n".cstring())
+      end
+      for d in data.values() do
+        twopc_reconnect_buffer.push(d)
+      end
+      []
+    end
+
+  fun ref throttled(conn: WallarooOutgoingNetworkActor ref) =>
+    if (not _throttled) or (not twopc_intro_done) then
+      _throttled = true
+      Backpressure.apply(_auth)
+      @printf[I32](("ConnectorSink is experiencing back pressure, " +
+        "connected = %s\n").cstring(), _connected.string().cstring())
+    end
+
+  fun ref unthrottled(conn: WallarooOutgoingNetworkActor ref) =>
+    if _throttled and twopc_intro_done then
+      _throttled = false
+      Backpressure.release(_auth)
+      @printf[I32](("ConnectorSink is no longer experiencing" +
+        " back pressure, connected = %s\n").cstring(),
+      _connected.string().cstring())
+      for d in twopc_reconnect_buffer.values() do
+        ifdef "checkpoint_trace" then
+          @printf[I32]("DBG: unthrottled: writing buffered %d bytes\n".cstring(), d.size())
+        end
+        try (conn as ConnectorSink ref)._write_final(d, None) else Fail() end
+      end
+      twopc_reconnect_buffer.clear()
+    end
+
+  fun send_msg(conn: WallarooOutgoingNetworkActor ref, msg: cp.Message) =>
+    let w1: Writer = w1.create()
```

Comment:
Can't we reuse an existing `Writer` that we hold onto as a field on this class?

Reply:
Sure, though these messages aren't frequent enough to worry about the GC penalty.  But this reminds me to investigate refactoring with `Bytes.length_encode()` instead.

Reply:
Two uses of `Bytes.length_encode()` were added by commit d3d3680310